protocol/test/core/ascend-zaps-test.ts

```diff
@@ -0,0 +1,557 @@
+            singleSwap,
+            fundMng,
+            0,
+            deadline
+          )
+      ).to.be.revertedWith('SwapDataDoesNotMatch');
+
+      singleSwap.assetIn = TEMPLE.address;
+
+      await ascendZaps
+        .connect(alan)
+        .exitVaultEarly(
+          swapAmount,
+          vault.address,
+          singleSwap,
+          fundMng,
+          0,
+          deadline
+        );
+    });
+
+    it('balancer swap fail', async () => {
+      singleSwap.assetOut = constants.tokens.USDC;
+      await vault.connect(alan).approve(ascendZaps.address, swapAmount);
+      await advanceTimeAndBlock(periodDuration);
+
+      await expect(
+        ascendZaps
+          .connect(alan)
+          .exitVaultEarly(
+            swapAmount,
+            vault.address,
+            singleSwap,
+            fundMng,
+            0,
+            deadline
+          )
+      ).to.be.revertedWith('BalancerVaultSwapError');
+
+      singleSwap.assetOut = constants.tokens.DAI;
+
+      await ascendZaps
+        .connect(alan)
+        .exitVaultEarly(
+          swapAmount,
+          vault.address,
+          singleSwap,
+          fundMng,
+          0,
+          deadline
+        );
+    });
+
+    it('insufficient temple balance', async () => {
+      const amount = toAtto(1000000).add(1);
+      await TEMPLE.mint(alan.address, amount);
+
+      await TEMPLE.connect(alan).approve(vault.address, amount);
+      await vault.connect(alan).deposit(amount);
+
+      swapAmount = amount;
+      singleSwap.amount = amount;
+
+      await vault.connect(alan).approve(ascendZaps.address, swapAmount);
+      await advanceTimeAndBlock(periodDuration);
+
+      await expect(
+        ascendZaps
+          .connect(alan)
+          .exitVaultEarly(
+            swapAmount,
+            vault.address,
+            singleSwap,
+            fundMng,
+            0,
+            deadline
+          )
+      ).to.be.revertedWith('InsufficientTempleBalance');
+    });
+
+    it('exit vault early swaps TEMPLE to DAI on balancer vault', async () => {
+      await vault.connect(alan).approve(ascendZaps.address, swapAmount);
+      await expect(
+        ascendZaps
+          .connect(alan)
+          .exitVaultEarly(
+            swapAmount,
+            vault.address,
+            singleSwap,
+            fundMng,
+            0,
+            deadline
+          )
+      ).to.be.revertedWith('FirstVaultCycle');
+
+      await advanceTimeAndBlock(periodDuration);
+      const res = await vault.inEnterExitWindow();
+      const cycleNum = res.cycleNumber;
+      expect(cycleNum).to.be.eq(1);
+
+      const beforeBalance = await TEMPLE.balanceOf(balanceVaultAddress);
+      const beforeDAIBalance = await DAI.balanceOf(alan.address);
+
+      const amountCalculated = await ascendZaps
+        .connect(alan)
+        .callStatic.exitVaultEarly(
+          swapAmount,
+          vault.address,
+          singleSwap,
+          fundMng,
+          0,
```

Comment:
get swap limits from balancer and verify user has at least that many additional tokens.

Reply:
done!